### PR TITLE
Transpose midi clips in song editor

### PR DESCRIPTION
**Pull request by allejok96**
Lets you quickly transpose a single track, section of the song or the whole song without messing around with master pitch. Right click a midi clip or selection:

image

Side note: Undo is wonky like every other multi-track action, but a suggested fix is coming later.

### CODE REVIEW

**Comment by allejok96** [comment]
Thanks @superpaik. The menu content depend on what clip you click.

Initially I thought it would be a good idea ignore beat clips (zero-length notes), because that would keep drums the same while transposing the melody. But I forgot to exclude beat clips from transposing when you right click on *another* clip... Now when I think about it, beat clips in song editor is a legacy thing so why give them special treatment? And I'm sure there are people who want to transpose beat clips too. So I'll add that option.

As for the other options I think Clear all notes could be completely removed from Song editor.

File: src/gui/MidiClipView.cpp
Comment (JohannesLorenz) [inline src/gui/MidiClipView.cpp:154]: QtCreator smashes a [warning](https://github.com/KDE/clazy/blob/1.11/docs/checks/README-range-loop-detach.md) here.

**Review by JohannesLorenz — CHANGES_REQUESTED** [review]
I know you plan more changes for beat clips, but here goes my review for the current change set.

**Comment by JohannesLorenz** [comment]
One minor thing from a user point of view: If the dialog would somehow tell you that you can only transpose upwards by 20 or downwards by `NumKeys-1-20`, this would give a better idea of what you *can* transpose, instead of saying: Here are `2*NumKeys-1` possible transposals, but many of them won't make sense. I imagine running a for loop over all selections and checking what is possible before displaying that dialogue would be fast enough.

**Comment by allejok96** [comment]
I'm inclined to keep the current behavior where the transpose option is hidden from beat clips, because otherwise when you right click a clip in Pattern editor it will show "Transpose" and that doesn't make much sense...

If you so happen to have a beat clip in Song editor and select all clips and transpose, there's no harm in letting that beat clip be transposed too... At least then you have an obscure (read LMMS-ish) way to transpose even those clips in Song editor, as opposed to excluding them from transpose completely.

Or what do you say?

File: src/gui/MidiClipView.cpp
Comment (JohannesLorenz) [reply src/gui/MidiClipView.cpp:154]: Actually I think this warning can not be fixed. I can't make the `clipview` `const`... Does that warning make any sense?

**Comment by JohannesLorenz** [comment]
> I'm inclined to keep the current behavior where the transpose option is hidden from beat clips

I don't have too much of an opinion. It's probably OK to hide it there, and if it's requested to show it later, we can still push a simple fix.

File: src/gui/MidiClipView.cpp
Comment (allejok96) [reply src/gui/MidiClipView.cpp:191]: Thanks! Never thought of that.

File: src/gui/MidiClipView.cpp
Comment (allejok96) [reply src/gui/MidiClipView.cpp]: Fixed. Now it checks max/min possible transpose.

**Comment by allejok96** [comment]
Transpose dialog only shows valid options now. No clipping.

I did this by adding `getBounds` and `transpose` to `NoteVector`. Maybe that's too Pythonic, but I figured it could be reused to reduce some code in PianoRoll in the future.

File: src/gui/MidiClipView.cpp
Comment (JohannesLorenz) [inline src/gui/MidiClipView.cpp:187]: Is that line equal to
```c++
update();
if( getGUI()->pianoRoll()->currentMidiClip() == m_clip )
{
	getGUI()->pianoRoll()->update();
}
```
? I really can't find out where that signal leads to. At least, the piano roll is being refreshed now.

File: src/gui/MidiClipView.cpp
Comment (JohannesLorenz) [inline src/gui/MidiClipView.cpp:168]: Of course, this now restricts to only do transposing that does not change the distance between the notes (i.e. does not bash all notes into one key). If that was wanted, ignore this comment.

**Review by JohannesLorenz — APPROVED** [review]
Code looks mostly good, see comments.

From testing point of view: What I find interesting is the behavior if I select only non-MidiClips and then do right-click on a MidiClip while the non-MidiClip selection is still active. It still allows to transpose the right-clicked MidiClip. I guess that right-clicked MidiClip gets part of the selection at time of right-clicking, so it's wanted?

File: src/gui/MidiClipView.cpp
Comment (JohannesLorenz) [reply src/gui/MidiClipView.cpp:168]: Oh right, just reading your comments that it was intended. Solving my own comment.

File: src/gui/MidiClipView.cpp
Comment (allejok96) [reply src/gui/MidiClipView.cpp:187]: Yes, it updates both the clip view and piano roll if active

File: src/core/Note.cpp
Comment (allejok96) [reply src/core/Note.cpp:284]: Can't hurt. Added check for `dataChanged()`. But by design there must be at least one clip with notes to show the transpose option and at least 1 semitone transpose before anything is performed so I left `setModified()` as is.

The journal issue (creating checkpoint for tracks where nothing changed) would be fixed by the Batch undo PR. The checkpoints would still be created, but restored as one.

**Comment by allejok96** [comment]
> I guess that right-clicked MidiClip gets part of the selection at time of right-clicking, so it's wanted?

Kinda correct. `getClickedClips()` returns the selection *or* the clicked clip if it's outside the selection. The behavior is the same when setting clip colors.

**Review by JohannesLorenz — APPROVED** [review]

File: src/gui/MidiClipView.cpp
Comment (DomClark) [inline src/gui/MidiClipView.cpp:168]: Personally, I would have written "Semitones to transpose *by*:", but I'll leave it up to you.

File: include/Note.h
Comment (allejok96) [reply include/Note.h:253]: So I should make them static methods of `NoteVector` instead?

Comment (DomClark) [reply include/Note.h:253]: I would leave `NoteVector` as a typedef for `QVector<Note*>`, and make them free (non-member) functions, so
```cpp
// See also my later comment for further changes to getBounds
bool getBounds(const NoteVector& notes, TimePos& start, TimePos& end, int& lower, int& upper);
bool transpose(const NoteVector& notes, int semitones);
```

Comment (allejok96) [reply include/Note.h:253]: I don't see a lot of these kind of functions in header files (I want at least `getBounds` to be accessible), so I figured putting it in a class would make things cleaner. Or is there some other way to not have a ton of functions in the global scope?

Comment (DomClark) [reply include/Note.h:253]: > Or is there some other way to not have a ton of functions in the global scope?

#6174?

> I don't see a lot of these kind of functions in header files (I want at least getBounds to be accessible), so I figured putting it in a class would make things cleaner.

I really wouldn't worry about it too much. You say you want to make the functions reusable - perhaps the reason there aren't many free functions in header files is because we haven't been very good at making things reusable up until now. If you look in namespace `std`, there are loads of free functions, but that doesn't cause anyone a problem. If you want, you can add `notes` or something to the names to make the purpose clearer without a class name for context (e.g. `boundsForNotes`, `transposeNotes`), but even without, you won't have problems with name collisions since parameter types distinguish functions in C++ (unlike in C, Python, etc.).

File: src/gui/MidiClipView.cpp
Comment (allejok96) [reply src/gui/MidiClipView.cpp]: You're the better English speaker!

File: src/core/Note.cpp
Comment (allejok96) [reply src/core/Note.cpp]: Thanks for the tip! I didn't know about std::optional. That's great.

File: src/core/Note.cpp
Comment (allejok96) [reply src/core/Note.cpp]: Removed

File: include/Note.h
Comment (allejok96) [reply include/Note.h]: Ok, that makes sense. I moved out `getBounds` and renamed it `boundsForNotes`. And `transpose` is so simple that I just dropped it.

**Comment by allejok96** [comment]
Fixed Dom's comments and rebased

File: src/core/Note.cpp
Comment (JohannesLorenz) [inline src/core/Note.cpp:262]: Just for my understanding: A negative side effect of returning `std::optional<NoteBounds>` instead of passing the four parameters via references (as "out parameters") is, using the return value might cause a copy? However, in C++17, it seems to be [forced to elide the copying](https://en.cppreference.com/w/cpp/language/copy_elision), which is why we do not need to worry?

**Review by JohannesLorenz — APPROVED** [review]
Code OK: Dom's fixes seems to be applied correctly. Testing also (still) works.

**Review by PhysSong — APPROVED** [review]
Ready to merge once you resolve conflicts.

File: src/core/Note.cpp
Comment (DomClark) [inline src/core/Note.cpp:248]: I wonder whether `return std::nullopt;` would be clearer here.

File: src/gui/clips/MidiClipView.cpp
Comment (DomClark) [inline src/gui/clips/MidiClipView.cpp:158]: `qobject_cast` would be more efficient here (and further down in this file).

**Review by DomClark — APPROVED** [review]
Only two minor comments - the code is fine as is, so feel free to ignore them.

File: src/core/Note.cpp
Comment (PhysSong) [reply src/core/Note.cpp:248]: I thought `{}` was enough, but `std::nullopt` looks clearer to me.

File: src/core/Note.cpp
Comment (allejok96) [reply src/core/Note.cpp:265]: Interesting. Even prior to C++17
> compilers are permitted, but not required to omit the copy

I did some tests. With `optional` it's a different story, because the value seems to be constructed first and then copied into the `optional` object. To avoid it you must do

    return std::optional<NoteBounds>{std::in_place, start, end, lower, upper};

I'm feeling tempted to remove all the `optional` fuzz and just go with `operator bool()`

File: src/gui/clips/MidiClipView.cpp
Comment (allejok96) [reply src/gui/clips/MidiClipView.cpp:158]: Thanks! I'll add it for the sake of good practice.

File: src/core/Note.cpp
Comment (allejok96) [reply src/core/Note.cpp]: I decided to drop `std::optional` and go with `operator bool()` to return true if the NoteBound is non-zero. That way we get return value optimization like @JohannesLorenz in a clean and simple way. Thoughts?

File: src/core/Note.cpp
Comment (DomClark) [reply src/core/Note.cpp:263]: At -O1 or above, GCC optimises out the extra copy and stores the constructor arguments straight into the optional. (See https://godbolt.org/z/aKoWcrGnY.) Even if it didn't, it's only copying 16 bytes in a hot memory region. An optional is better semantically than giving a special meaning to a particular value of `NoteBounds`, and I wouldn't sacrifice that unless profiling showed it to be especially problematic.

Comment (allejok96) [reply src/core/Note.cpp:263]: Agreed that this is not a problem, disagreeing on semantics. But that's why you're the C++ guy and I'm not (cool site btw)

**Comment by JohannesLorenz** [comment]
3 reviews and all rework done. Merge?

**Pull request merged**
